File: Problems/Leetcode/3001-4000/3001-3100/3083.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool isSubstringPresent(string s) {
        // bool vis[26][26] {};
        int vis[26] {};
        for (int i = 1; i < s.size(); i++) {
            int x = s[i - 1] - 'a', y = s[i] - 'a';
            // vis[x][y] = true;
            // if (vis[y][x]) {
            //     return true;
            // }

            //* 位运算优化
            vis[x] |= 1 << y;
            if (vis[y] >> x & 1) {
                return true;
            }
        }
        return false;
    }
};

File: Problems/Leetcode/3001-4000/3001-3100/3083.py
from itertools import pairwise


class Solution:
    def isSubstringPresent(self, s: str) -> bool:
        """
        ss = s[::-1]
        for i in range(len(s) - 1):
            if s[i : i + 2] in ss:
                return True
        return False
        """
        st = set()
        for x, y in pairwise(s):
            st.add((x, y))
            if (y, x) in st:
                return True
        return False